File: Interview_Questions/543_diameterOfBinaryTree.py
'''
Intuition: The maximum path can be one of two cases for any binary tree
1) Connecting left and right path through the tree root
  (left height - 1) + (right height - 1) // -1 on each to rule out the root
2) The path of some subtree down the line

Clarification on 1)
        1
      2   3
      left height = 2
      right height = 2
      total path is 2 = (2-1) + (2-1)

Example of case 2)
        1
    2       3
          4   5
        6       7
      8
'''
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        resultTuple = self.helper(root)
        return max(resultTuple[0]-1, resultTuple[1])
            
    def helper(self, root: Optional[TreeNode]) -> List[int]:
        leftStats, rightStats = [1,0], [1,0]
        currentBranch = 0 
        # I have a left height
        if root.left is not None:
            leftStats = self.helper(root.left)
            leftStats[0] += 1
            currentBranch += leftStats[0] - 1
        # I have a right height
        if root.right is not None:
            rightStats = self.helper(root.right)
            rightStats[0] += 1
            currentBranch += rightStats[0] - 1

        # Find my biggest height
        maxHeight = max(leftStats[0], rightStats[0])
        
        # Find my biggest branch
        maxBranch = max(
            currentBranch, # current branch (LH + RH - 1 - 1)
            max(leftStats[1], rightStats[1])# biggest branch I've seen
        )
        return [maxHeight, maxBranch]

File: DFS/543_DiameterOfBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
     if (root == nullptr) return 0;
        
        int maxPath = 0;
        helpMaxDiameter(root, maxPath);

        return maxPath;
    }
    
    int helpMaxDiameter(TreeNode* node, int& maxPath) {
        if (node->left == nullptr && node->right == nullptr) return 0;

        int leftPathLength = 0;
        if (node->left != nullptr) 
            leftPathLength = helpMaxDiameter(node->left, maxPath) + 1;
        int rightPathLength = 0;
        if (node->right != nullptr) 
            rightPathLength = helpMaxDiameter(node->right, maxPath) + 1;

        int fullPath = leftPathLength + rightPathLength;
        maxPath = max(maxPath, fullPath);

        return max(leftPathLength, rightPathLength);
    }
};
